### DISCUSSION TITLE
ROS2 Design: Retrieval of data from a topic

### DISCUSSION THREAD

**Discussion by Xindictus** [gstavrinos/ai_msc_robotic_systems_assignments#35] [Q&A] [answered]
If we need to dynamically retrieve specific information like camera details (e.g., focal lengths) or other data just once within our Node, we ideally aim to separate our concerns and uphold SOLID principles, just like how it goes with software in general. Within the context of ROS2, what high-level architecture would be most suitable for this scenario?

- _**Service call**_
Define a custom service which subscribes to the camera info topic and responds with the required information to incoming requests.
- _**Classic subscription callback**_
Just subscribe to the topic in the same Node, attach a callback and grab what we need.
- **_Something else_**
???

**Comment by gstavrinos** [comment, marked as answer]
A service client would be ideal if the information you need is provided through a service.

In the case of information being only available through a topic, the most optimal way is to create a subscriber, and once you acquire the information you need, [destroy the subscription](https://docs.ros2.org/latest/api/rclpy/api/node.html#rclpy.node.Node.destroy_subscription).
